rasa_core/policies/embedding_policy.py

```diff
@@ -0,0 +1,1022 @@
+
+            sim_act = tf.reduce_sum(emb_act[:, :, 0:1, :] *
+                                    emb_act[:, :, 1:, :], -1)
+            sim_act *= tf.expand_dims(mask, 2)
+
+            return sim, sim_act
+        else:
+            raise ValueError("Wrong similarity type {}, "
+                             "should be 'cosine' or 'inner'"
+                             "".format(self.similarity_type))
+
+    def _tf_loss(self, sim, sim_act, mask):
+        """Define loss"""
+
+        if self.use_max_sim_neg:
+            max_sim_neg = tf.reduce_max(sim[:, :, 1:], -1)
+            loss = (tf.maximum(0., self.mu_pos - sim[:, :, 0]) +
+                    tf.maximum(0., self.mu_neg + max_sim_neg)) * mask
+
+        else:
+            # create an array for mu
+            mu = self.mu_neg * np.ones(self.num_neg + 1)
+            mu[0] = self.mu_pos
+            mu = mu[np.newaxis, np.newaxis, :]
+
+            factors = tf.concat([-1 * tf.ones([1, 1, 1]),
+                                 tf.ones([1, 1, tf.shape(sim)[-1] - 1])],
+                                axis=-1)
+
+            max_margin = tf.maximum(0., mu + factors * sim)
+
+            loss = tf.reduce_sum(max_margin, -1) * mask
+
+        # penalize max similarity between intent embeddings
+        loss_act = tf.maximum(0., tf.reduce_max(sim_act, -1))
+        loss += loss_act * self.C_emb
+
+        loss = tf.reduce_sum(loss, -1) / tf.reduce_sum(mask, 1)
+        # add regularization losses
+        loss = tf.reduce_mean(loss) + tf.losses.get_regularization_loss()
+        return loss
+
+    def _create_tf_graph(self, a_in, b_in, c_in):
+        """Create tf graph for training"""
+
+        if self.share_embedding:
+            name_a = 'a_and_b'
+            name_b = 'a_and_b'
+        else:
+            name_a = 'a'
+            name_b = 'b'
+
+        a = self._create_tf_nn(a_in,
+                               self.num_hidden_layers_a,
+                               self.hidden_layer_size_a,
+                               self.droprate['a'],
+                               name=name_a)
+        a = tf.layers.dropout(a, rate=self.droprate['a'],
+                              training=self.is_training)
+        emb_utter = self._create_embed(a, name=name_a)
+
+        b = self._create_tf_nn(b_in,
+                               self.num_hidden_layers_b,
+                               self.hidden_layer_size_b,
+                               self.droprate['b'],
+                               name=name_b)
+        shape_b = tf.shape(b)
+        b = tf.layers.dropout(b, rate=self.droprate['b'],
+                              training=self.is_training,
+                              noise_shape=[shape_b[0],
+                                           shape_b[1],
+                                           1, shape_b[-1]])
+        emb_act = self._create_embed(b, name=name_b)
+
+        c = c_in
+        # TODO do we need hidden layers for slots?
```

Comment:
@Ghostvv do we?

Reply:
i don't think so, i set it as a reminder for myself